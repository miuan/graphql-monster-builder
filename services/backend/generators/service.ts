import {
    SchemaModel,
    SchemaModelRelationType,
    StructureBackend,
    SchemaModelMember,
    SchemaModelType,
} from '../../common/types'

import { writeToFile, templateFileToText } from '../../common/files'
import { getOnlyOneRelatedMember, firstToLower } from '../../common/utils'
import { relatedParamName1Id, relatedParamName2Id, relatedParamName1, relatedParamName2 } from './schema'

import logger from '../../log'
import { BackendDirectory } from '../backendDirectory'
const log = logger.getLogger('service')

const defaultMembers = ['createdAt', 'updatedAt', 'id']

const memberCreateAndRemoveLinks = (model: SchemaModel, member: SchemaModelMember) => {
    const modelName = model.modelName
    const relation = member.relation
    const relatedMember = getOnlyOneRelatedMember(member)

    const ret = {
        result: '',
        connect: '',
    }

    if (!relatedMember) {
        return ret
    }

    const lower = firstToLower(modelName)
    const relationName = relation.name
    const funcAddToName = `addTo${relationName}`
    const funcRemoveFromName = `removeFrom${relationName}`
    const relatedModelName = member.relation.relatedModel.modelName

    ret.result = templateFileToText('service-add-remove.ts', {
        _LOWER_NAME_: lower,
        _RELATION_NAME_: relationName,
        _RELATED_PARAM_NAME_1_: relatedParamName1Id(model, relatedMember),
        _RELATED_PARAM_NAME_2_: relatedParamName2Id(member),
        _RELATED_MEMBER_NAME_: relatedMember.name,
        _MEMBER_NAME_: member.name,
        _LOWER_RELATED_MODEL_NAME_: firstToLower(relatedModelName),
        _RELATED_MODEL_NAME_: relatedModelName,
        _PAYLOAD_PARAM_1: relatedParamName1(model, relatedMember),
        _PAYLOAD_PARAM_2: relatedParamName2(member),
    })

    ret.connect += `${funcAddToName} : ${funcAddToName}(entry),\n${funcRemoveFromName} : ${funcRemoveFromName}(entry),`

    return ret
}

const modelCreateAddRemoveLinks = (model: SchemaModel, membersWirhRelation: SchemaModelMember[]) => {
    const ret = {
        result: '',
        connect: '',
    }

    for (const member of membersWirhRelation) {
        const { result, connect } = memberCreateAndRemoveLinks(model, member)

        ret.result += result
        ret.connect += connect
    }
    return ret
}

export const createService = (model: SchemaModel) => {
    const modelName = model.modelName
    const lower = firstToLower(modelName)
    const varName = lower + 'Model'

    const membersWithRelation = model.members.filter(
        (model) => model.relation && model.relation.type === SchemaModelRelationType.RELATION,
    )

    const allIdsConversionsCreate = conversionsIdsToField(membersWithRelation, true)
    const allIdsConversionsUpdate = conversionsIdsToField(membersWithRelation, false)
    const connectRelationCreate = updateLinkedModels(membersWithRelation, 'createdModel.id')
    const connectRelationUpdate = updateLinkedModels(membersWithRelation, 'updatedModel.id')
    const disconnectRelations = disconnectLinkedModels(modelName, membersWithRelation)
    const disconnectRelationsInRemove = disconnectLinkedModels(modelName, membersWithRelation, true)
    const { result: serviceAddRemove, connect: serviceAddRemoveConnect } = modelCreateAddRemoveLinks(
        model,
        membersWithRelation,
    )

    let actionBeforeCreate = ''
    let actionAfterCreate = ''
    let actionBeforeUpdate = ''
    if (modelName === 'User') {
        actionBeforeCreate = `
    if(data.password) {
      data.__password = await extras.generateHash(data.password);
      data.password = '******';
    }
`
        actionAfterCreate = ``
        actionBeforeUpdate = 'extras.checkPasswordIsNotIncluded(data);'
    }

    let result = templateFileToText(`service.ts`, {
        _MODEL_NAME_: modelName,
        _LOWER_NAME_: lower,
        VAR_NAME: varName,
        _ALL_IDS_CONVERSIONS_CREATE_: allIdsConversionsCreate,
        _ALL_IDS_CONVERSIONS_UPDATE_: allIdsConversionsUpdate,
        _CONNECT_RELATION_CREATE_: connectRelationCreate,
        _CONNECT_RELATION_UPDATE_: connectRelationUpdate,

        _DISCONNECT_RELATIONS_IN_REMOVE: disconnectRelationsInRemove,
        _DISCONNECT_RELATIONS_: disconnectRelations,

        _EXTRA_ACTION_BEFORE_CREATE_: actionBeforeCreate,
        _EXTRA_ACTION_AFTER_CREATE_: actionAfterCreate,
        _EXTRA_ACTION_BEFORE_UPDATE_: actionBeforeUpdate,
        _EXTRA_ACTION_AFTER_UPDATE_: '',
        _SERVICE_ADD_REMOVE_CONNECT_: serviceAddRemoveConnect,
        _SERVICE_ADD_REMOVE_: serviceAddRemove,
    })
    result += ''
    return result
}

export const disconnectLinkedModels = (modelName, membersWithRelation: SchemaModelMember[], isItForRemove = false) => {
    let result = ''
    // const membersWithRelation = members.every(m => m.relation != null);

    for (const member of membersWithRelation) {
        const relation = member.relation
        const relatedModel = relation.relatedModel
        const relationModelName = relatedModel.modelName
        const relationName = relation.name
        const relatedMember = relatedModel.members.find((m) => m.relation && m.relation.name === relationName)
        const relatedMemberName = relatedMember.name

        const dataMemberName = `data.${member.name}`
        const lower = relationModelName.charAt(0).toLowerCase() + relationModelName.slice(1)
        let removeList

        result += ``

        // remove doesn't have any incoming data, should disconnect them automaticaly
        // but can happend some relations are already removed bacause is called from there
        // example: M2 have required relation to M1, M1 is removing and also call remove M2 object
        //          M2 will clean all its own relations, but cleaning relations to M1 is not necessay
        //          because M1 it will removed by itself anyway
        if (isItForRemove) {
            result += `if( !skipRelations.includes('${lower}') ){`
            removeList = `[...skipRelations, '${modelName}']`
        } else {
            removeList = `['${modelName}']`
            if (member.isArray) {
                result += `if( (data.${relation.payloadNameForId} && data.${relation.payloadNameForId}.length > 0) || (data.${relation.payloadNameForCreate} && data.${relation.payloadNameForCreate}.length > 0) ){`
            } else {
                result += `if( data.${relation.payloadNameForId} || data.${relation.payloadNameForCreate} ){`
            }
        }

        if (relatedMember.isArray) {
            // more about `$all` see https://stackoverflow.com/a/18149149
            result += `
      // relation is type: ${relation.type.toString()} 
      await entry.models['${lower}'].updateMany({${relatedMemberName}:{$all: [id]}}, {$pull: {${relatedMemberName}: id}})
    `
        } else if (relatedMember.isRequired) {
            result += `
      // relation is type: ${relation.type.toString()} and related model have it as required
      const relatedModels = await entry.models['${lower}'].find({${relatedMemberName}: id}, {_id:true})
      if(relatedModels && relatedModels.length > 0){
        // we call services because is necessary also unlink relations with removing object
        await Promise.all(relatedModels.map((relatedModel) => entry.services['${lower}'].remove(relatedModel._id, ctxUserId, ${removeList})))
      }
  `
        } else {
            result += `
      // relation is type: ${relation.type.toString()}
      await entry.models['${lower}'].updateMany({${relatedMemberName}: id}, {${relatedMemberName}: null})
  `
        }

        result += '}'
    }

    return result
}

export const updateLinkedModels = (membersWithRelation: SchemaModelMember[], currentIdName, create = false) => {
    let result = ''
    // const membersWithRelation = members.every(m => m.relation != null);

    for (const member of membersWithRelation) {
        const relationModelName = member.relation.relatedModel.modelName
        const relatedModel = member.relation.relatedModel
        const relationName = member.relation.name
        const relatedMember = relatedModel.members.find((m) => m.relation && m.relation.name === relationName)
        const relatedMemberName = relatedMember.name
        const varLinkedIds = `${member.name}LinkedIds`

        const lower = relationModelName.charAt(0).toLowerCase() + relationModelName.slice(1)

        result += `
    if(${varLinkedIds} && ${varLinkedIds}.length > 0) {`
        // TODO: test for relatedMember.relation.type generate correct `updateMany`
        // TODO: test for relatedMember.relation.type set right type corespond to member model
        if (relatedMember.isArray) {
            result += `
        await entry.models['${lower}'].updateMany({ _id: {$in: ${varLinkedIds}} }, {  $push: {${relatedMemberName}: { $each: [${currentIdName}]}} })
      `
        } else {
            result += `
      await entry.models['${lower}'].updateMany({ _id: {$in: ${varLinkedIds}} }, {  ${relatedMemberName}: ${currentIdName} })
    `
        }
        result += '}'
    }

    return result
}

export const conversionsIdsToField = (membersWithRelation: SchemaModelMember[], create = false) => {
    let result = ''
    // const membersWithRelation = members.every(m => m.relation != null);

    for (const member of membersWithRelation) {
        const isMeMany = member.isArray
        const relationModelName = member.relation.relatedModel.modelName
        const varLinkedIds = `${member.name}LinkedIds`
        const lower = relationModelName.charAt(0).toLowerCase() + relationModelName.slice(1)

        const relatedConnecteMember = member.relation.relatedModel.members.find(
            (member) => member.relation && member.relation.name === member.relation.name,
        )
        const isHeMany = relatedConnecteMember.isArray

        let swithcOfLInkedIds
        let switchOfAddConnectedId = ''
        if (create) {
            if (relatedConnecteMember.isRequired) {
                switchOfAddConnectedId =
                    '// the related member is required so we need for creation a id (TEMPORARY-ID)\n\t\t'
                swithcOfLInkedIds =
                    '// backward relation is setup just with TEMPORARY-ID so need update later with REAL-ID\n\t\t'
            } else {
                switchOfAddConnectedId =
                    '// the related member is NOT required so we will update later with REAL-ID\n\t\t// '
                swithcOfLInkedIds = '// backward relation is not setup yet, so need update later with REAL-ID\n\t\t '
            }
        } else {
            swithcOfLInkedIds = '// backward relation is already setup, so no need any update aditional\n\t\t// '
        }

        const transformIds = templateFileToText(
            isMeMany ? 'service-transform-many-ids.ts' : 'service-transform-one-id.ts',
            {
                _LOWER_NAME_: lower,
                _LINKDED_IDS_: varLinkedIds,
                _CONNECTED_MEMBER_NAME_: relatedConnecteMember.relation.payloadNameForId,
                _CONNECTED_MEMBER_ID_: isHeMany ? '[id]' : 'id',
                _MEMBER_NAME_: member.name,
                _SWITCH_OF_ADD_CONNECTED_ID_: switchOfAddConnectedId,
                _SWITCH_OF_ADD_TO_LINKED_IDS_: swithcOfLInkedIds,
                _PAYLOAD_NAME_FOR_ID_: member.relation.payloadNameForId,
                _PAYLOAD_NAME_FOR_CREATE_: member.relation.payloadNameForCreate,
            },
        )

        if (isMeMany) {
            result += `
        ${transformIds}
      `
        } else {
            result += `
        ${transformIds}
      `
        }
    }

    return result
}

function createEntityService(model: SchemaModel) {
    const result = templateFileToText('entity.service.t.ts', {
        __ENTITY_NAME__: model.modelName,
    })

    return result
}

export const generateServiceToFile = (backendDirectory: BackendDirectory, model: SchemaModel) => {
    const str = model.type === SchemaModelType.MODEL ? createService(model) : createEntityService(model)

    backendDirectory.servicesWrite(`${model.modelName}`, str)
}

export const generateServices = (backendDirectory: BackendDirectory, models: SchemaModel[]) => {
    log.trace('generateServices')
    for (const model of models) {
        log.info(`Generate service for model: ${model.modelName}`)
        generateServiceToFile(backendDirectory, model)
    }
}
