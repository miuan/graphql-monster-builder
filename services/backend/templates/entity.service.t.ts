/**
 * Generated by GraphQL Monster Builder (GMB) version: __VERSION__
 *     Do NOT edit! For more info see: https://github.com/miuan/graphql-monster-builder
 */
export const generateCreate__ENTITY_NAME__ = (entry) => (entity) => {

}
export const generateUpdate__ENTITY_NAME__ = (entry) => (entity) => {

}

export const generateRemove__ENTITY_NAME__ = (entry) => (entity) => {

}

export function generate__ENTITY_NAME__Service(entry){
    return {
        create: generateCreate__ENTITY_NAME__(entry),
        update: generateUpdate__ENTITY_NAME__(entry),
        remove: generateRemove__ENTITY_NAME__(entry)
    }
}