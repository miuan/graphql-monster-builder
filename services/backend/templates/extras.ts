import * as bcrypt from 'bcrypt-nodejs';
import * as jwt from 'jsonwebtoken';
import * as _ from 'lodash'

export const generateHash = function (password) {
  return bcrypt.hashSync(password, bcrypt.genSaltSync(8), null);
};

/**
 *
 * @param user
 * @param options
 */
export const generateTokenJWT = (tokenizeData, opts?): string => {
  let options = opts || {};

  if (!options) {
    options = {
      /* expires in 365 days */
      expiresIn: '365d',
    };
  }

  const tokenJWT = jwt.sign(tokenizeData, process.env.JWT_TOKEN_SECRET || 'protectql_test_secret', options);

  return tokenJWT;
};

export const genPasswordAndTokens = (userData) => {
  // TODO: schort the expire date
  const options = { expiresIn: '365d' };
  userData.token = generateTokenJWT({ id: userData.id, role: userData.role }, options);
  userData.refreshToken = generateTokenJWT({ id: userData.id }, options);
};


export const checkPasswordIsNotIncluded = (userData) => {
  if (userData.password) {
    throw 'User can change password only in userChangePasswordMutation(old, new)';
  }
};

export const generateLogin = (entry) => {
  return async (root, data, ctx) => {
    
    const user = await entry.models['user'].findOne({ email: data.email });

    if (user && bcrypt.compareSync(data.password, user._password)) {
      return user;
    }

    throw 'Email or password is not correct';
  };
};

export const generateParentLogin = (entry) => async (ctx) => {
  if(ctx.params.parentAccessToken != process.env.PARENT_ACCESS_TOKEN
    || ctx.params.parentUserId != process.env.PARENT_ACCESS_USER_ID){
    throw 'Unknown parent access token'
  }
  
  const adminEmail = process.env.ADMIN_EMAIL

  const user = await entry.models['user'].findOne({ email: adminEmail })
  if(!user.token) {
    genPasswordAndTokens(user)
    await user.save()
  }
  ctx.body = {token: user.token}
}

export const generateChangePassword = (entry) => {
  return async (root, data, ctx) => {
    const userModel = entry.models['user'];
    const user = await userModel.findById(data.userId);

    console.log(data, user);
    if (!user) {
      throw `Unknown user with id: ${data.userId}`;
    }

    if (!bcrypt.compareSync(data.oldPassword, user._password)) {
      throw `Old password is different than is currently used`;
    }

    const updatedUser = await userModel.findByIdAndUpdate(
      data.userId, 
      { _password:generateHash(data.newPassword) }, 
      { new: true },
    );

    return updatedUser;
  };
};

export const generateProtections = (entry, modelName) => {
  return {
    public: () => true,
    user: (ctx) => {
      if (ctx && ctx.state && ctx.state.user) {
        return true;
      }
    },
    owner: async (ctx, data, ownerField = 'user') => {
      if (ctx && ctx.state && ctx.state.user) {
        // NOTE: in edit, if `fd` contain also other fields, what they are edited
        //       then `await entry.models[modelName].findOne(fd)` not find anything
        //       because data are not corespond with what is in DB
        // const fd = {
        //   ...data
        // }
        // if(fd.id) {
        //   fd._id = fd.id
        //   delete fd.id
        // }
        // const u = await entry.models[modelName].findOne(fd, ownerField)
        
        const u = await entry.models[modelName].findById(data.id || data._id, ownerField)
        console.log('OWNER CHECK : user1 >>', {u, fd:(data.id || data._id), state: ctx.state, modelName, models: entry.models})
        console.log('u && u[ownerField] == ctx.state.user.id', u && u[ownerField] == ctx.state.user.id)
        return u && u[ownerField] == ctx.state.user.id;
      } else {
        return false
      }
    },
    role: async (ctx, roles: string[]) => {
      if (ctx && ctx.state && ctx.state.user) {
        const ctxUser = ctx.state.user;
        const userRoleModel = entry.models['userRole'];
      
        const userRole = await userRoleModel.findOne({ role: roles[0], users: ctxUser.id });
      
        return userRole != null;
      }
    },
    filter: async (ctx, data, filters: any[], roles: string[]) => {
      if (ctx && ctx.state && ctx.state.user) {
        const ctxUser = ctx.state.user;
        const dataFilter = data.filter || {}
 
        // if is something restricted to filter can't be filter behind OR
        // because it will pass security 
        // but show even data what is under protection
        if(dataFilter.OR) {
          return false
        }

        // TODO: INPORTANT! is catch if secure filter is behind one $and example: filter: {$and:[{user: {{userId}}}]}
        //       but not catch if is under depper $and ... example filter: {$and:[{$and:[{user: {{userId}}}]]}
        let AND = null
        if(dataFilter.AND) {
          AND = dataFilter.AND
        } else if(dataFilter.length) {
          AND = dataFilter
        } else {
          AND = [dataFilter]
        }

        for(let a of AND) {
          for(let filter of filters) {
            if(filter.name) {
              const value = _.get(a, filter.name)
              console.log('filter', filter.name, value)

              if(filter.value == value || (filter.value == '{{userId}}' && value == ctxUser.id)) {
                return true
              }
              // console.log('data.filter.AND', data.filter.AND)
              // console.log('data.filter.AND[0]', data.filter.AND[0])
              // console.log('data.filter.AND[0].user_every', data.filter.AND[0].user_every)
            }
          }
        }
        
      }
      return false
    },
    checkDataContainProtectedFields
  };
};

// ******************************
// * * *    F I L T E R     * * *
// ******************************

const filterANDOR = (filter, operator) => {
  if(filter.length == 1){
    return filterGen(filter[0])
  }
  else if(filter.length > 1) {
    const obj = {}
    obj[operator] = []
    for(const f of filter) {
      obj[operator].push(filterGen(f))
    }
    return obj
  } else {
    return undefined
  }

}


export const filterGen = (filter) => {

  console.log('DEBUG:FILTER', filter)
  let obj = {}

  if(!filter){
    return obj
  } else if(filter.AND) {
    return filterANDOR(filter.AND, '$and')
  } else if(filter.OR) {
    return filterANDOR(filter.OR, '$or')
  } 

  if(Object.keys(filter).length > 0) {
    for(const f in filter){
      const ff = filter[f]

      if(/_every$/.test(f)){
        obj[f.substr(0, f.length-6)] = filterGen(ff)
      } else if(/_none$/.test(f)){
        obj[f.substr(0, f.length-5)] = {'$not': filterGen(ff)}
      } else if(/_not_contains$/.test(f)){
        obj[f.substr(0, f.length-13)] = {$not: { "$regex": ff, "$options": "i" }}
      } else if(/_contains$/.test(f)){
        obj[f.substr(0, f.length-9)] = { "$regex": ff, "$options": "i" }
      } else if(/_not$/.test(f)){
        obj[f.substr(0, f.length-4)] = { $ne: ff }
      } else if(/_not_in$/.test(f)){
        obj[f.substr(0, f.length-7)] = { $not: { $in: ff } }
      } else if(/_in$/.test(f)){
        obj[f.substr(0, f.length-3)] = { $in: ff }
      } else if(/_not_starts_with$/.test(f)){
        obj[f.substr(0, f.length-16)] = {$not: { "$regex": `^${ff}`, "$options": "i" }}
      } else if(/_starts_with$/.test(f)){
        obj[f.substr(0, f.length-12)] = { "$regex": `^${ff}`, "$options": "i" }
      } else if(/_not_ends_with$/.test(f)){
        obj[f.substr(0, f.length-14)] = {$not: { "$regex": `${ff}$`, "$options": "i" }}
      } else if(/_ends_with$/.test(f)){
        obj[f.substr(0, f.length-10)] = { "$regex": `${ff}$`, "$options": "i" }
      } else if(/_lt$/.test(f)){
        obj[f.substr(0, f.length-3)] = { "$lt": ff }
      } else if(/_lt$/.test(f)){
        obj[f.substr(0, f.length-3)] = { "$lt": ff }
      } else if(/_lte$/.test(f)){
        obj[f.substr(0, f.length-4)] = { "$lte": ff }
      } else if(/_gt$/.test(f)){
        obj[f.substr(0, f.length-3)] = { "$gt": ff }
      } else if(/_gte$/.test(f)){
        obj[f.substr(0, f.length-4)] = { "$gte": ff }
      } else if(f == 'id') {
        obj['_id'] = ff
      } else if(typeof ff == 'string' || typeof ff == 'number') {
        obj = {...filter}
      } else {
        // TODO: stop going deeper than just object.id
        //       mongoose can't query populated fields
        //       https://stackoverflow.com/questions/17535587/mongoose-query-a-populated-field
        //       posible show message is funciton of prepaid version
        return filterGen(ff)
      }
    }
  } else {
    return filter
  }
  
  return obj
}

class ReachProtectedFields extends Error {
  fields: any

  constructor(protectedFields) {
    super(`Reach protected fields`)
    this.fields = protectedFields
  }
}

/**
 * should return any field with name starts double underscore example  `__port`, `__readolny`
 * 
 * @param data - data to check
 * @param path - actual path in data
 */
export const checkDataContainProtectedFields = (data, path='/') => {

  let founded = []
  const keys = Object.keys(data)
  if(keys.length > 0){
    for(const fieldName of keys){
      if(/^__/.test(fieldName) || /^_/.test(fieldName)){
        founded.push({name: fieldName, path: path})
      } else if(typeof(data[fieldName]) != 'string') {
        let fundedInDeep = checkDataContainProtectedFields(data[fieldName], `${path}${fieldName}/`)
        founded.push(...fundedInDeep)
      }
    }
  }

  return founded
}

export const createUser = async (entry, email: string, password: string, rolesIds: string[] = []) => {
  const userModel = entry.models['user']
  const userService = entry.services['user']

  let user = await userModel.findOne({email: email})

  if(!user) {
    // call service instead of simple model
    // because we want also create relation between user <- and -> roles
    user = await userService.create({email: email, _password: password, rolesIds})
    console.log(`User with email: ${email} and with pass: ${password} [CREATED]`, user)
  } else {
    user = await userService.update({_password: password, rolesIds}, user.id)
    console.log(`Update pass (${password}) to already existed user with email: ${email}`, user)
  }

  return user
}

export const createRole = async (entry, role: string) => {
  const userRoleModel = entry.models['userRole']
  let userRole = await userRoleModel.findOne({role})
  if(!userRole){
    userRole = await userRoleModel.create({role})
    console.log(`Role with name: ${role} [CREATED]`, userRole)
  } else {
    console.log(`Role with name: ${role} already exist`, userRole)
  }
  
  return userRole
}