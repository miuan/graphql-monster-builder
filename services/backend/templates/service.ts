/* eslint-disable prettier/prettier */
import { VAR_NAME, Types } from '../models/_MODEL_NAME_'
import * as extras from '../extras'

export const _LOWER_NAME_All = (entry) => {
    return async (data) => {
        if (entry.hooks && entry.hooks.services['before_MODEL_NAME_All']) {
            await entry.hooks.services['before_MODEL_NAME_All'](entry, data)
        }

        const filter = extras.filterGen(data.filter)
        let findPromise = _LOWER_NAME_Model.find(filter)
        if (data.skip) findPromise = findPromise.skip(data.skip)
        if (data.limit) findPromise = findPromise.limit(data.limit)
        if (data.orderBy) {
            const {
                groups: { field, type },
            } = data.orderBy.match(/(?<field>\w+)_(?<type>(asc|desc))/)
            findPromise = findPromise.sort([[field, type]])
        }

        let models = await findPromise.lean()

        if (entry.hooks && entry.hooks.services['after_MODEL_NAME_All']) {
            models = await entry.hooks.services['after_MODEL_NAME_All'](entry, { models, ...data })
        }
        return models
    }
}

export const _LOWER_NAME_Count = (entry) => {
    return async (data) => {
        if (entry.hooks && entry.hooks.services['before_MODEL_NAME_Count']) {
            await entry.hooks.services['before_MODEL_NAME_Count'](entry, data)
        }

        const filter = extras.filterGen(data.filter)
        let countPromise = _LOWER_NAME_Model.count(filter)

        let count = await countPromise.lean()

        if (entry.hooks && entry.hooks.services['after_MODEL_NAME_Count']) {
            count = await entry.hooks.services['after_MODEL_NAME_Count'](entry, { count, ...data })
        }
        return count
    }
}

export const _LOWER_NAME_One = (entry) => {
    return async (id, userId = null) => {
        if (entry.hooks && entry.hooks.services['before_MODEL_NAME_One']) {
            await entry.hooks.services['before_MODEL_NAME_One'](entry, { id })
        }
        let model = await _LOWER_NAME_Model.findById(id).lean()
        model.id = id

        if (entry.hooks && entry.hooks.services['after_MODEL_NAME_One']) {
            model = await entry.hooks.services['after_MODEL_NAME_One'](entry, { id, model })
        }

        return model
    }
}

export const _LOWER_NAME_Create = (entry) => {
    return async (data, ctxUserId = null) => {
        // before real object exist
        // we generate TEMPORARY-ID for related objects what they have a required relation...
        const id = Types.ObjectId()

        _ALL_IDS_CONVERSIONS_CREATE_
        if (entry.hooks && entry.hooks.services['before_MODEL_NAME_Create']) {
            data = await entry.hooks.services['before_MODEL_NAME_Create'](entry, { data, ctxUserId })
        }
        _EXTRA_ACTION_BEFORE_CREATE_
        let createdModel = await VAR_NAME.create(data)

        _EXTRA_ACTION_AFTER_CREATE_
        _CONNECT_RELATION_CREATE_
        if (entry.hooks && entry.hooks.services['after_MODEL_NAME_Create']) {
            createdModel = await entry.hooks.services['after_MODEL_NAME_Create'](entry, {
                id: createdModel._id,
                data: createdModel,
                ctxUserId,
            })
        }
        return createdModel
    }
}

export const _LOWER_NAME_Update = (entry) => {
    return async (data, update_LOWER_NAME_Id = null, ctxUserId = null) => {
        let id = update_LOWER_NAME_Id

        if (data.id) {
            id = data.id
            delete data.id
        }

        if (entry.hooks && entry.hooks.services['before_MODEL_NAME_Update']) {
            data = await entry.hooks.services['before_MODEL_NAME_Update'](entry, { data, id, ctxUserId })
        }

        // disconnect all relations
        _DISCONNECT_RELATIONS_
        _ALL_IDS_CONVERSIONS_UPDATE_
        _EXTRA_ACTION_BEFORE_UPDATE_
        let updatedModel = await VAR_NAME.findByIdAndUpdate(id, data, { new: true, runValidators: true })

        // connect all relations
        _CONNECT_RELATION_UPDATE_

        if (entry.hooks && entry.hooks.services['after_MODEL_NAME_Update']) {
            updatedModel = await entry.hooks.services['after_MODEL_NAME_Update'](entry, {
                data: updatedModel,
                id,
                ctxUserId,
            })
        }

        return updatedModel
    }
}

export const _LOWER_NAME_Remove = (entry, ctxUserId = null) => {
    return async (id, userId, skipRelations = []) => {
        if (entry.hooks && entry.hooks.services['before_MODEL_NAME_Remove']) {
            await entry.hooks.services['before_MODEL_NAME_Remove'](entry, { id, ctxUserId })
        }

        // disconnect all relations
        _DISCONNECT_RELATIONS_IN_REMOVE
        _EXTRA_ACTION_BEFORE_REMOVE_
        let removedModel = await VAR_NAME.findByIdAndRemove(id)

        if (entry.hooks && entry.hooks.services['after_MODEL_NAME_Remove']) {
            removedModel = await entry.hooks.services['after_MODEL_NAME_Remove'](entry, {
                data: removedModel,
                id,
                ctxUserId,
            })
        }

        return removedModel
    }
}

_SERVICE_ADD_REMOVE_

export const generate_MODEL_NAME_Service = (entry) => {
    return {
        all: _LOWER_NAME_All(entry),
        count: _LOWER_NAME_Count(entry),
        one: _LOWER_NAME_One(entry),
        create: _LOWER_NAME_Create(entry),
        update: _LOWER_NAME_Update(entry),
        remove: _LOWER_NAME_Remove(entry),
        _SERVICE_ADD_REMOVE_CONNECT_
    }
}
