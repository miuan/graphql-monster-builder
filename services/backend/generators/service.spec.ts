
import 'jest-extended'
import * as fs from 'fs'
import { SchemaModel, SchemaModelProtection, SchemaModelProtectionType } from '../../common/types'
import * as service from './service'
import * as extras from '../templates/extras'
import * as _ from 'lodash'

const TEST_DIR = 'templates/__generated_services_for_test__'
const SERVICE_BACKEND = './services/backend/'
const testDirFullPath = `${SERVICE_BACKEND}${TEST_DIR}`

const createService = (model) => {

    const created = service.createService(model)
    const camelName = _.camelCase(model.modelName)
    fs.writeFileSync(`${testDirFullPath}/services/${camelName}Service.ts`, created)
    fs.writeFileSync(`${testDirFullPath}/models/${camelName}.ts`, `export const ${camelName}Model = {
        create: jest.fn(), //.mockImplementation(async (data)=>data),
        findByIdAndUpdate: jest.fn(), //.mockImplementation(async (id, data)=>data),
        findByIdAndRemove: jest.fn() //.mockImplementation(async (data)=>data),
    } as any;
    export default ${camelName}Model`)

    try{
        const service = require(`../${TEST_DIR}/services/${camelName}Service.ts`)
        const model = require(`../${TEST_DIR}/models/${camelName}.ts`)[`${camelName}Model`]
        return {service, model, mockClearToAll: () => {
            model.create.mockClear()
            model.findByIdAndUpdate.mockClear()
            model.findByIdAndRemove.mockClear()
        }}
    } catch (ex) {
        // console.log(ex)
        throw ex
    }
    
}

const createEntryMock = (resolverName:string, method: string) => {
    const lowerMethod = method.toLowerCase()
    const entry = {
        hooks: {},
        services: {} as any
    }
    entry.services[resolverName]={}
    entry.services[resolverName][lowerMethod] = jest.fn().mockResolvedValue(`${lowerMethod}_output`)

    // entry.hooks[`beforetestResolver${method}`] = jest.fn().mockResolvedValue({beforeHookResponse: `${lowerMethod}_output`})
    // entry.hooks[`aftertestResolver${method}`] = jest.fn().mockResolvedValue({afterHookResponse: `${lowerMethod}_output`})

    return entry
}

describe('service', () => {
    let existsSyncMock
    let writeFileSyncMock

    let testResolverWithPublic
    beforeAll(()=>{
        
        if(!fs.existsSync(testDirFullPath)){
            fs.mkdirSync(testDirFullPath)
        } 
        
        if(!fs.existsSync(`${testDirFullPath}/services`)){
            fs.mkdirSync(`${testDirFullPath}/services`)
        } 

        if(!fs.existsSync(`${testDirFullPath}/models`)){
            fs.mkdirSync(`${testDirFullPath}/models`)
        } 

        //if(!fs.existsSync(`${testDirFullPath}/extras.ts`)){
            fs.writeFileSync(`${testDirFullPath}/extras.ts`, `
export const filterGen = jest.fn()
export default filterGen
`)
        //}
        
    })



    afterAll(()=>{
        fs.rmdirSync(testDirFullPath)
    })


    describe('hooks', ()=>{
        let hookService
        beforeAll(()=>{
            hookService = createService({modelName:'Hook', members:[]})

        })

        describe.each([
            'before',
            'after'
        ])('%s ', (method)=>{

            beforeEach(()=>{
                hookService.mockClearToAll()
            })

            it.each([
                ['create', 'create'], 
                ['update', 'findByIdAndUpdate'], 
                ['remove', 'findByIdAndRemove']
            ])('%s', async (forWho, modelMethod)=>{
                const hookName = _.camelCase(`${method} Hook ${forWho}`)
                const methodName = _.camelCase(`hook ${forWho}`)
                const entry = { hooks: {services : {}}}
                entry.hooks.services[hookName] = jest.fn()

                
                
                const methodFn = hookService.service[methodName](entry)

                const aId = 'random-id-value'
                const aObject = {}
                const aResultedObject = {_id:aId}

                hookService.model[modelMethod].mockImplementation(()=>aResultedObject)
                
                if(forWho == 'create') await methodFn(aObject)
                else if(forWho == 'update') await methodFn(aObject, aId)
                else if(forWho == 'remove') await methodFn(aId, aObject)

                expect(entry.hooks.services[hookName]).toBeCalledTimes(1)
                expect(entry.hooks.services[hookName].mock.calls[0][0]).toEqual(entry)

                // hook create doesn't have id
                if(hookName!=='beforeHookCreate') {
                    expect(entry.hooks.services[hookName].mock.calls[0][1]).toHaveProperty('id', aId)
                }

                // hook before remove doesn't have data
                if(hookName!=='beforeHookRemove') {
                    const hookData = method == 'before' ? aObject : aResultedObject
                    expect(entry.hooks.services[hookName].mock.calls[0][1]).toHaveProperty('data', hookData)
                }

                expect(hookService.model[modelMethod]).toBeCalledTimes(1)
                

        })
    })
})
    

})