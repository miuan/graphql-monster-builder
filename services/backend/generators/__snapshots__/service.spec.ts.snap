// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`service hooks after  create 1`] = `
Object {
  "data": undefined,
}
`;

exports[`service hooks after  remove 1`] = `
Object {
  "data": undefined,
  "id": Object {},
  "userId": null,
}
`;

exports[`service hooks after  update 1`] = `
Object {
  "data": undefined,
  "id": null,
}
`;

exports[`service hooks before  create 1`] = `
Object {
  "data": Object {},
}
`;

exports[`service hooks before  remove 1`] = `
Object {
  "id": Object {},
}
`;

exports[`service hooks before  update 1`] = `
Object {
  "data": Object {},
  "id": null,
}
`;
